### ISSUE TITLE
Installation issue: llvm+clang from manylinux2014 has non-functional clang

### ISSUE THREAD

**Issue by robertu94** [LLNL/spack#48073]
### Steps to reproduce the issue

```console
$ spack mirror add develop-2024-12-08-developer-tools-manylinux2014 https://binaries.spack.io/develop-2024-12-08/developer-tools-manylinux2014
$ spack install llvm+clang
$ clang
...
```

### Error message

<details><summary>Error message</summary>

<pre>
clang: error: '/opt/rh/devtoolset-10/root/usr/lib/gcc/x86_64-redhat-linux/10' does not contain a GCC installation
clang: error: no input files
</pre></details>

### Information on your system

* **Spack:** 0.24.0.dev0 (94bd7b9afb8f905a443644c55b1cea2f786d7077)
* **Python:** 3.13.0
* **Platform:** linux-fedora41-icelake

### Additional information

The clang installed via the build cache expects to find the rhel-devtoolset that is not installed on most machines.  It is likely not relocated and/or packaged correctly.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by haampie** [comment]
That's currently out of scope to relocate because the gcc dependency is not modelled.

You can remove the `--gcc-install-dir` flag from `bin/*.cfg`, which may result in llvm picking up a gcc in some system path.

**Comment by robertu94** [comment]
You're right you can remove it, and it fixes clang and clang++, but it does not fix clang-tidy and similar which still produces this error.  Alternatively, you can modify the `.cfg` files to contain: ```--gcc-install-dir=$spack_gcc_prefix/lib/gcc/$gcc_target_triple/$gcc_version``` where gcc_version is something like `14.2.0` and `gcc_target_triple` is something like `x86_64-pc-linux-gnu`

**Comment by haampie** [comment]
it depends on the llvm version

**Comment by robertu94** [comment]
Another way this manifests is the difference in the abi of std::string used by the RHEL devtoolset.  It causes missing symbols when linking to llvm built in this way.  For example this function is missing `_ZN5clang12ast_matchers8internal14HasNameMatcherC1ESt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS9_EE` which decodes to `clang::ast_matchers::internal::HasNameMatcher::HasNameMatcher(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)` 

We should remove the `llvm` package from the buildcaches until we fix or find a better solution to these problems. There are few things more frustrating than a binary package that breaks things.

cc the other LLVM maintainers @skosukhin @trws

**Comment by haampie** [comment]
That's a separate issue, the solution is in #48065.

It's better to focus on a specific problem in a github issue.

**Comment by trws** [comment]
@haampie, given that #48065 builds its own gcc and libstdc++, is this not also solved by that?

**Comment by haampie** [comment]
The original issue of paths to gcc not being relocated is still there, the collapsed bits are resolved yes.

For `llvm@:15` this path is hard-coded into the binaries and not easy to fix until ^gcc is a non-external dependency.

For `llvm@16:` there is a config file, so the workaround is to modify that manually.

#45189 may still not entirely fix this since ^gcc would likely be an external.
